src/dymaptic.GeoBlazor.Core/Objects/AttributesDictionary.cs

```diff
@@ -0,0 +1,313 @@
+    /// </summary>
+    /// <param name="key">
+    ///    The key to remove
+    /// </param>
+    public async Task Remove(string key)
+    {
+        _backingDictionary.Remove(key);
+
+        await OnChange.InvokeAsync();
+    }
+
+    /// <summary>
+    ///     Removes all keys and values from the dictionary.
+    /// </summary>
+    public async Task Clear()
+    {
+        _backingDictionary.Clear();
+
+        await OnChange.InvokeAsync();
+    }
+
+    /// <summary>
+    ///     Determines whether the dictionary contains the specified key.
+    /// </summary>
+    /// <param name="key">
+    ///     The key to locate in the dictionary
+    /// </param>
+    /// <returns>
+    ///     True if the key was found, false otherwise.
+    /// </returns>
+    public bool ContainsKey(string key)
+    {
+        return _backingDictionary.ContainsKey(key);
+    }
+
+    /// <summary>
+    ///     Determines whether the dictionary contains the specified value.
+    /// </summary>
+    /// <param name="value">
+    ///     The value to locate in the dictionary
+    /// </param>
+    /// <returns>
+    ///     True if the value was found, false otherwise.
+    /// </returns>
+    public bool ContainsValue(object value)
+    {
+        return _backingDictionary.ContainsValue(value);
+    }
+
+    /// <inheritdoc />
+    public override bool Equals(object? obj)
+    {
+        if (ReferenceEquals(null, obj)) return false;
+        if (ReferenceEquals(this, obj)) return true;
+        if (obj.GetType() != GetType()) return false;
+
+        return Equals((AttributesDictionary)obj);
+    }
+
+    /// <summary>
+    ///   Determines whether the dictionary contains the same keys and values as the specified dictionary.
+    /// </summary>
+    public bool Equals(Dictionary<string, object> otherDictionary)
+    {
+        if (otherDictionary.Count != Count) return false;
+
+        foreach (KeyValuePair<string, object> pair in _backingDictionary)
+        {
+            if (!otherDictionary.TryGetValue(pair.Key, out object? value)) return false;
+
+            if (!value.Equals(pair.Value)) return false;
+        }
+
+        return true;
+    }
+
+    /// <inheritdoc />
+    public override int GetHashCode()
+    {
+        return base.GetHashCode();
+    }
+
+    internal AttributeSerializationRecord[] ToSerializationRecord()
+    {
+        return _backingDictionary
+            .Select(kvp =>
+                new AttributeSerializationRecord(kvp.Key, kvp.Value.ToString(),
+                    kvp.Value.GetType().ToString()))
+            .ToArray();
+    }
+
+    private readonly Dictionary<string, object> _backingDictionary;
+
+    /// <summary>
+    ///    Gets or the value associated with the specified key.
+    /// </summary>
+    /// <param name="key">
+    ///     The key to get the value for
+    /// </param>
+    public object this[string key] => _backingDictionary[key];
+}
+
+[ProtoContract(Name = "Attribute")]
```

Comment:
Had to do some research on protobuf to get a little better understanding.  Is speed the primary reason for serializing with protobuf in this situation?

Reply:
Yes, if you look at the protobuf serialization, you get highly-compressed data, where each class/property name is replaced by a number based on the position. This means less data per graphic, which equals to faster transfer when you have a large set.